internal/controllers/topology/cluster/desired_state.go

```diff
@@ -776,19 +783,20 @@ func computeMachineDeploymentVersion(s *scope.Scope, machineDeploymentTopology c
 	// Return early if the currentVersion is already equal to the desiredVersion
 	// no further checks required.
 	if currentVersion == desiredVersion {
-		return currentVersion, nil
+		return currentVersion
 	}
 
 	// Return early if the upgrade for the MachineDeployment is deferred.
 	if isMachineDeploymentDeferred(s.Blueprint.Topology, machineDeploymentTopology) {
-		s.UpgradeTracker.MachineDeployments.MarkDeferredUpgrade(currentMDState.Object.Name)
-		return currentVersion, nil
+		s.UpgradeTracker.MachineDeployments.MarkDeferredUpgrade(desiredMDName)
+		s.UpgradeTracker.MachineDeployments.MarkPendingUpgrade(desiredMDName)
+		return currentVersion
 	}
 
 	// Return early if we are not allowed to upgrade the machine deployment.
 	if !s.UpgradeTracker.MachineDeployments.AllowUpgrade() {
```

Comment:
Let's refactor AllowUpgrade.
We should just store something that AfterControlPlaneUpgrade is blocking in the tracker
and then we should have a separate check which checks for the upgrade concurrency. 

It's confusing that we mix it and I always have to look up what AllowUpgrade is doing. The name is way too generic to quickly understand what's going on.

This also allows us to not create new MachineDeployments when the AfterControlPlaneUpgrade hook blocks.